packages/core/src/navigation/navigation.element.ts

```diff
@@ -0,0 +1,479 @@
+  };
+
+  private toggle() {
+    this.expandedChange.emit(!this.expanded);
+  }
+
+  get currentActiveItem() {
+    return this.visibleChildren.find(c => c.id === this.ariaActiveDescendant);
+  }
+
+  protected get footerTemplate() {
+    return this.navigationEnd
+      ? html`
+          <div role="region" class="navigation-end">
+            <slot name="cds-navigation-end"></slot>
+          </div>
+        `
+      : '';
+  }
+
+  protected get startTemplate() {
+    let returnHTML;
+
+    this.rootNavigationStart
+      ? (returnHTML = html`
+          <div role="region" class="navigation-start" cds-layout="vertical">
+            <slot name="navigation-start"></slot>
+            <cds-divider class="start-divider"></cds-divider>
+          </div>
+        `)
+      : (returnHTML = '');
+
+    return returnHTML;
+  }
+
+  protected get visibleChildren(): FocusableElement[] {
+    return Array.from(this.allNavigationElements).filter(n => isVisible(n));
+  }
+
+  firstUpdated() {
+    if (this.rootNavigationStart) {
+      this.rootNavigationStart.addEventListener('click', () => {
+        this.toggle();
+      });
+    }
+
+    // set all visible navigation elements to tabindex = -1
+    this.allNavigationElements.forEach(item => {
+      setAttributes(item, ['tabindex', '-1']);
+    });
+  }
+
+  disconnectedCallback() {
+    super.disconnectedCallback();
+    this.removeEventListener('focus', this.initAriaActiveDescendant);
+    this.removeEventListener('keydown', this.keyboardNavigationHandler);
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+    this.tabIndex = 0;
+    this.addEventListener('focus', this.initAriaActiveDescendant);
+    this.addEventListener('keydown', this.keyboardNavigationHandler);
+    this.addEventListener('blur', this.blurActiveElement);
+    this.addEventListener('expandedChange', event => {
+      if (this.currentActiveItem && event.target instanceof CdsNavigationGroup) {
+        setFocus(this.currentActiveItem);
+      }
+    });
+  }
+
+  render() {
+    return html`<nav
+      tabindex="-1"
+      class="private-host"
+      aria-label="${this.i18n.navigationLabel}"
+      cds-layout="${DEFAULT_NAVIGATION_LAYOUT}
+                  wrap:none
+                  align:horizontal-center"
+    >
+      ${this.startTemplate}
+      <slot name="cds-navigation-substart"></slot>
+      <div class="navigation-body-wrapper" role="list" cds-layout="p-y:xxs">
+        <div
+          class="navigation-body"
+          role="presentation"
+          cds-layout="${this.layout}
+                    wrap:none"
+        >
+          <slot name="navigation-body"></slot>
+        </div>
+      </div>
+      ${this.footerTemplate}
+    </nav>`;
+  }
+
+  updated(props: PropertyValues<this>) {
+    super.updated(props);
+
+    if (props.has('expanded')) {
+      this.expandedRoot = this.expanded;
+    }
+
+    this.updateChildrenProps();
+  }
+
+  updateChildrenProps() {
+    if (this.navigationGroupItems) {
```

Comment:
There's a lot of repeated work here. It might be nice to have a utility that does the mechanics of these code blocks in the `lit-element.ts` helpers. Something like...

`syncPropsForAllItems<T>(items: T[], source: Element, conditions: { [prop: string]: boolean }`

So then these calls would all look like...

`syncPropsForAllItems(this.navigationGroupItems, this, { groupItem: true });`

...and you can move the logic for testing for length or truthyness into the helper fn.

It would also decrease the conditional branches we need to cover in this component method.

Reply:
ty for this idea I added what is essentially a wrapper for syncProps but can iterate on the array of items.